<a href="https://colab.research.google.com/github/krishshah8000/mlproject/blob/master/ML_Breast_cancer.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [ ]:
from sklearn.datasets import load_breast_cancer
import pandas as pd

# Load the dataset
breast_cancer = load_breast_cancer()

# Separate features (x) and target (y)
x = pd.DataFrame(breast_cancer.data, columns=breast_cancer.feature_names)
y = pd.DataFrame(breast_cancer.target, columns=['target'])

print("Features (x):")
display(x.head())

print("\nTarget (y):")
display(y.head())

Features (x):


,mean radius,mean texture,mean perimeter,mean area,mean smoothness,mean compactness,mean concavity,mean concave points,mean symmetry,mean fractal dimension,...,worst radius,worst texture,worst perimeter,worst area,worst smoothness,worst compactness,worst concavity,worst concave points,worst symmetry,worst fractal dimension
0,17.99,10.38,122.80,1001.0,0.11840,0.27760,0.3001,0.14710,0.2419,0.07871,...,25.38,17.33,184.60,2019.0,0.1622,0.6656,0.7119,0.2654,0.4601,0.11890
1,20.57,17.77,132.90,1326.0,0.08474,0.07864,0.0869,0.07017,0.1812,0.05667,...,24.99,23.41,158.80,1956.0,0.1238,0.1866,0.2416,0.1860,0.2750,0.08902
2,19.69,21.25,130.00,1203.0,0.10960,0.15990,0.1974,0.12790,0.2069,0.05999,...,23.57,25.53,152.50,1709.0,0.1444,0.4245,0.4504,0.2430,0.3613,0.08758
3,11.42,20.38,77.58,386.1,0.14250,0.28390,0.2414,0.10520,0.2597,0.09744,...,14.91,26.50,98.87,567.7,0.2098,0.8663,0.6869,0.2575,0.6638,0.17300
4,20.29,14.34,135.10,1297.0,0.10030,0.13280,0.1980,0.10430,0.1809,0.05883,...,22.54,16.67,152.20,1575.0,0.1374,0.2050,0.4000,0.1625,0.2364,0.07678



Target (y):


,target
0,0
1,0
2,0
3,0
4,0


In [ ]:
x.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 569 entries, 0 to 568
Data columns (total 30 columns):
 #   Column                   Non-Null Count  Dtype  
---  ------                   --------------  -----  
 0   mean radius              569 non-null    float64
 1   mean texture             569 non-null    float64
 2   mean perimeter           569 non-null    float64
 3   mean area                569 non-null    float64
 4   mean smoothness          569 non-null    float64
 5   mean compactness         569 non-null    float64
 6   mean concavity           569 non-null    float64
 7   mean concave points      569 non-null    float64
 8   mean symmetry            569 non-null    float64
 9   mean fractal dimension   569 non-null    float64
 10  radius error             569 non-null    float64
 11  texture error            569 non-null    float64
 12  perimeter error          569 non-null    float64
 13  area error               569 non-null    float64
 14  smoothness error         5

# Task
Prepare the breast cancer dataset for modeling by splitting it into training and testing sets and then standardizing the features using `StandardScaler`. After preparation, train a Logistic Regression model and simulate predictions to compare Mean Squared Error (MSE) and Binary Cross Entropy (BCE) loss values, interpreting which loss function penalizes incorrect predictions more severely.

## Prepare Data for Modeling

### Subtask:
Apply train-test split to the features (x) and target (y) and then standardize the features using StandardScaler to prepare them for the Logistic Regression model.


**Reasoning**:
The subtask requires splitting the data into training and testing sets and then standardizing the features using StandardScaler. I will import the necessary functions, perform the train-test split, instantiate and fit the scaler on the training data, and then transform both training and test data.



In [ ]:
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler

# Split the data into training and testing sets
x_train, x_test, y_train, y_test = train_test_split(x, y, test_size=0.2, random_state=42)

print("Shape of x_train:", x_train.shape)
print("Shape of x_test:", x_test.shape)
print("Shape of y_train:", y_train.shape)
print("Shape of y_test:", y_test.shape)

# Instantiate StandardScaler
scaler = StandardScaler()

# Fit on training data and transform both training and test data
x_train = scaler.fit_transform(x_train)
x_test = scaler.transform(x_test)

print("\nFeatures (x_train) after standardization (first 5 rows):\n", x_train[:5])
print("\nFeatures (x_test) after standardization (first 5 rows):\n", x_test[:5])

Shape of x_train: (455, 30)
Shape of x_test: (114, 30)
Shape of y_train: (455, 1)
Shape of y_test: (114, 1)

Features (x_train) after standardization (first 5 rows):
 [[-1.44075296 -0.43531947 -1.36208497 -1.1391179   0.78057331  0.71892128
   2.82313451 -0.11914956  1.09266219  2.45817261 -0.26380039 -0.01605246
  -0.47041357 -0.47476088  0.83836493  3.25102691  8.43893667  3.39198733
   2.62116574  2.06120787 -1.23286131 -0.47630949 -1.24792009 -0.97396758
   0.72289445  1.18673232  4.67282796  0.9320124   2.09724217  1.88645014]
 [ 1.97409619  1.73302577  2.09167167  1.85197292  1.319843    3.42627493
   2.01311199  2.66503199  2.1270036   1.55839569  0.80531919 -0.81268678
   0.75195659  0.87716951 -0.89605315  1.18122247  0.18362761  0.60059598
  -0.31771686  0.52963649  2.17331385  1.3112795   2.08161691  2.1374055
   0.76192793  3.26560084  1.92862053  2.6989469   1.89116053  2.49783848]
 [-1.39998202 -1.24962228 -1.34520926 -1.10978518 -1.33264483 -0.30735463
  -0.36555756 -0.6

## Train Logistic Regression Model

### Subtask:
Train a Logistic Regression model on the standardized training data. Although we are simulating predictions later, training a model is a necessary step to adhere to the request's flow.


**Reasoning**:
To train a Logistic Regression model, I need to import the class, instantiate it, and then fit it to the training data. I will make sure to ravel the y_train data as suggested.



In [ ]:
from sklearn.linear_model import LogisticRegression

# Instantiate Logistic Regression model
model = LogisticRegression(random_state=42)

# Train the model
model.fit(x_train, y_train.values.ravel())

print("Logistic Regression model trained successfully.")

Logistic Regression model trained successfully.


## Simulate Predicted Probabilities

### Subtask:
Create a dummy true target (y_true) and two sets of predicted probabilities: one for a highly correct prediction (0.9) and another for a highly incorrect prediction (0.1). This will allow us to evaluate the loss functions.


**Reasoning**:
To create the dummy true target and predicted probabilities as specified in the subtask, I will define these variables in a code block.



In [ ]:
y_true = 1
y_pred_correct = 0.9
y_pred_incorrect = 0.1

print(f"True target (y_true): {y_true}")
print(f"Correct prediction (y_pred_correct): {y_pred_correct}")
print(f"Incorrect prediction (y_pred_incorrect): {y_pred_incorrect}")

True target (y_true): 1
Correct prediction (y_pred_correct): 0.9
Incorrect prediction (y_pred_incorrect): 0.1


## Compute and Compare Loss Values

### Subtask:
Calculate the Mean Squared Error (MSE) and Binary Cross Entropy (BCE) loss for both the correct and incorrect simulated predictions. The results will then be printed.


**Reasoning**:
I need to import the required libraries, calculate MSE and BCE for both correct and incorrect predictions, and then print the results as per the subtask instructions.



In [ ]:
from sklearn.metrics import mean_squared_error
import tensorflow as tf

# Calculate MSE for correct prediction
mse_correct = mean_squared_error([y_true], [y_pred_correct])

# Calculate MSE for incorrect prediction
mse_incorrect = mean_squared_error([y_true], [y_pred_incorrect])

# Instantiate BinaryCrossentropy loss function
bce = tf.keras.losses.BinaryCrossentropy(from_logits=False)

# Calculate BCE for correct prediction
bce_correct = bce([y_true], [y_pred_correct]).numpy()

# Calculate BCE for incorrect prediction
bce_incorrect = bce([y_true], [y_pred_incorrect]).numpy()

print(f"Mean Squared Error (Correct Prediction): {mse_correct:.4f}")
print(f"Mean Squared Error (Incorrect Prediction): {mse_incorrect:.4f}")
print(f"Binary Cross Entropy (Correct Prediction): {bce_correct:.4f}")
print(f"Binary Cross Entropy (Incorrect Prediction): {bce_incorrect:.4f}")

Mean Squared Error (Correct Prediction): 0.0100
Mean Squared Error (Incorrect Prediction): 0.8100
Binary Cross Entropy (Correct Prediction): 0.1054
Binary Cross Entropy (Incorrect Prediction): 2.3026


# Task
## Evaluate Model Performance

### Subtask:
Compute and display classification performance metrics (e.g., accuracy, precision, recall, F1-score, confusion matrix) for the trained Logistic Regression model on the test set (x_test, y_test).

## Compute Performance Metrics

### Subtask:
Calculate and display classification performance metrics (e.g., accuracy, precision, recall, F1-score, confusion matrix) for the trained Logistic Regression model on the test set (x_test, y_test).


**Reasoning**:
I need to make predictions using the trained Logistic Regression model on the test set (`x_test`) and store them as `y_pred` to prepare for calculating the performance metrics.



In [ ]:
y_pred = model.predict(x_test)

print("Predictions (y_pred) on the test set (first 5 rows):\n", y_pred[:5])

Predictions (y_pred) on the test set (first 5 rows):
 [1 0 0 1 1]


**Reasoning**:
To calculate and display classification performance metrics, I will import the required functions from `sklearn.metrics` and then compute the accuracy, precision, recall, F1-score, and confusion matrix using `y_test` and `y_pred`, finally printing all the results.



In [ ]:
from sklearn.metrics import accuracy_score, precision_score, recall_score, f1_score, confusion_matrix

# Calculate Accuracy
accuracy = accuracy_score(y_test, y_pred)

# Calculate Precision
precision = precision_score(y_test, y_pred)

# Calculate Recall
recall = recall_score(y_test, y_pred)

# Calculate F1-score
f1 = f1_score(y_test, y_pred)

# Calculate Confusion Matrix
conf_matrix = confusion_matrix(y_test, y_pred)

print(f"Accuracy: {accuracy:.4f}")
print(f"Precision: {precision:.4f}")
print(f"Recall: {recall:.4f}")
print(f"F1-score: {f1:.4f}")
print("Confusion Matrix:\n", conf_matrix)

Accuracy: 0.9737
Precision: 0.9722
Recall: 0.9859
F1-score: 0.9790
Confusion Matrix:
 [[41  2]
 [ 1 70]]


# Task
Perform K-Fold Cross-Validation and Bootstrap Sampling to evaluate the Logistic Regression model on the breast cancer dataset, comparing the mean and standard deviation of accuracy across both methods.

## Perform K-Fold Cross-Validation

### Subtask:
Apply K-Fold Cross-Validation to the trained Logistic Regression model using the training data (`x_train`, `y_train`). Calculate and display the mean and standard deviation of accuracy across the folds.


**Reasoning**:
To apply K-Fold Cross-Validation as per the instructions, I will import the necessary classes and functions, instantiate KFold, and then use cross_val_score to get the accuracy scores across the folds. Finally, I will calculate and print the mean and standard deviation of these scores.



In [ ]:
from sklearn.model_selection import KFold, cross_val_score
import numpy as np

# Instantiate KFold
kf = KFold(n_splits=5, shuffle=True, random_state=42)

# Perform K-Fold Cross-Validation
cv_scores = cross_val_score(model, x_train, y_train.values.ravel(), cv=kf, scoring='accuracy')

# Calculate mean and standard deviation of accuracy
mean_accuracy = np.mean(cv_scores)
std_accuracy = np.std(cv_scores)

print(f"K-Fold Cross-Validation Accuracy (Mean): {mean_accuracy:.4f}")
print(f"K-Fold Cross-Validation Accuracy (Standard Deviation): {std_accuracy:.4f}")

K-Fold Cross-Validation Accuracy (Mean): 0.9780
K-Fold Cross-Validation Accuracy (Standard Deviation): 0.0120


## Perform Bootstrap Sampling

### Subtask:
Apply Bootstrap Sampling to evaluate the trained Logistic Regression model. Calculate and display the mean and standard deviation of accuracy across multiple bootstrap samples.

## Perform Bootstrap Sampling

### Subtask:
Apply Bootstrap Sampling to evaluate the trained Logistic Regression model. Calculate and display the mean and standard deviation of accuracy across multiple bootstrap samples.

#### Instructions
1. Initialize an empty list to store bootstrap accuracy scores.
2. Define the number of bootstrap iterations (e.g., `n_iterations = 100`).
3. Inside a loop for the specified number of iterations:
   a. Generate a bootstrap sample: Randomly sample with replacement from the training data indices to create `x_bootstrap` and `y_bootstrap`.
   b. Train a new Logistic Regression model on the `x_bootstrap` and `y_bootstrap` data. (Use `random_state` for reproducibility if desired, but not strictly necessary for bootstrap).
   c. Evaluate the newly trained model on the `x_test` and `y_test` data to get the accuracy score.
   d. Append the accuracy score to the list.
4. Calculate the mean and standard deviation of the collected bootstrap accuracy scores.
5. Print the mean and standard deviation of the Bootstrap Sampling accuracy, formatted to a few decimal places.

**Reasoning**:
Now, I will implement the Bootstrap Sampling as outlined in the previous instruction block. This involves setting up a loop to create bootstrap samples, training a model on each sample, evaluating it on the test set, and then calculating the mean and standard deviation of the accuracy scores.



In [ ]:
from sklearn.utils import resample

# Initialize list to store bootstrap accuracy scores
bootstrap_accuracies = []

# Define the number of bootstrap iterations
n_iterations = 100

for i in range(n_iterations):
    # Generate a bootstrap sample of indices from the training data
    # x_train.shape[0] gives the number of samples in x_train
    bootstrap_indices = resample(np.arange(x_train.shape[0]), replace=True, n_samples=x_train.shape[0], random_state=i)

    # Create bootstrap samples for x and y
    x_bootstrap = x_train[bootstrap_indices]
    y_bootstrap = y_train.iloc[bootstrap_indices].values.ravel()

    # Train a new Logistic Regression model on the bootstrap sample
    bootstrap_model = LogisticRegression(random_state=42) # Using a fixed random_state for model training for consistency
    bootstrap_model.fit(x_bootstrap, y_bootstrap)

    # Evaluate the model on the original test set
    y_pred_bootstrap = bootstrap_model.predict(x_test)
    accuracy_bootstrap = accuracy_score(y_test, y_pred_bootstrap)

    # Append the accuracy score
    bootstrap_accuracies.append(accuracy_bootstrap)

# Calculate mean and standard deviation of bootstrap accuracies
mean_bootstrap_accuracy = np.mean(bootstrap_accuracies)
std_bootstrap_accuracy = np.std(bootstrap_accuracies)

print(f"Bootstrap Sampling Accuracy (Mean): {mean_bootstrap_accuracy:.4f}")
print(f"Bootstrap Sampling Accuracy (Standard Deviation): {std_bootstrap_accuracy:.4f}")

Bootstrap Sampling Accuracy (Mean): 0.9776
Bootstrap Sampling Accuracy (Standard Deviation): 0.0069


## Compare K-Fold and Bootstrap Results

### Subtask:
Compare the summary statistics (mean and standard deviation) of the performance metrics obtained from both K-Fold Cross-Validation and Bootstrap Sampling. Discuss the differences, advantages, and disadvantages of each method for this specific model and dataset.


#### Comparison of K-Fold Cross-Validation and Bootstrap Sampling Results

**K-Fold Cross-Validation:**
*   Mean Accuracy: 0.9780
*   Standard Deviation of Accuracy: 0.0120

**Bootstrap Sampling:**
*   Mean Accuracy: 0.9776
*   Standard Deviation of Accuracy: 0.0069

**Discussion:**

1.  **Comparison of Results:**
    *   The mean accuracy from both methods is very similar (0.9780 for K-Fold and 0.9776 for Bootstrap), suggesting that on average, the Logistic Regression model performs consistently well on this dataset, regardless of the evaluation method.
    *   The standard deviation of accuracy from Bootstrap Sampling (0.0069) is lower than that from K-Fold Cross-Validation (0.0120). This indicates that the performance estimates from Bootstrap Sampling are slightly less variable, meaning the accuracy scores obtained across different bootstrap samples were more consistent with each other compared to the accuracy scores across different K-folds.

2.  **Advantages of K-Fold Cross-Validation:**
    *   **Efficient Use of Data:** Every data point is used for training and testing exactly once across the folds, which can lead to a less biased estimate of the model's true performance.
    *   **Reduced Variance (compared to a single train-test split):** By averaging results over multiple folds, it reduces the variability of the performance estimate, making it more reliable than a single split.
    *   **Good for Limited Data:** Ensures that all data contribute to both training and validation, which is beneficial when dataset size is a concern.

3.  **Advantages of Bootstrap Sampling:**
    *   **Robustness to Small Datasets:** It can be particularly useful for smaller datasets where splitting into train/test sets might leave very little data for training or testing. Bootstrap samples effectively increase the apparent sample size for evaluation.
    *   **Estimates of Estimator Statistics:** Bootstrap not only provides a performance estimate but also a distribution of these estimates, which allows for calculation of confidence intervals for model performance, offering more insight into the stability of the model.
    *   **Flexibility:** It can be adapted to estimate the variability of almost any statistic (not just accuracy).

4.  **Disadvantages:**
    *   **K-Fold Cross-Validation:** While reducing bias, it can still have higher variance in performance estimates compared to bootstrap for very small datasets. Also, the choice of 'k' can impact the results.
    *   **Bootstrap Sampling:** Each bootstrap sample is drawn with replacement, meaning some original data points are not included in a sample (out-of-bag samples), and some are included multiple times. This introduces an artificial bias in the sample compared to the original dataset. Also, it can be computationally more intensive if a large number of iterations are used.

## Final Task

### Subtask:
Provide a comprehensive comparison of k-fold cross-validation and bootstrap sampling results, highlighting their implications for model evaluation.


## Summary:

### Q&A
The comprehensive comparison of k-fold cross-validation and bootstrap sampling results highlights that both methods yielded very similar mean accuracy for the Logistic Regression model on the breast cancer dataset, with K-Fold Cross-Validation showing a mean accuracy of 0.9780 and Bootstrap Sampling a mean accuracy of 0.9776. However, Bootstrap Sampling demonstrated a lower standard deviation of accuracy (0.0069) compared to K-Fold Cross-Validation (0.0120), indicating greater consistency in its performance estimates for this specific scenario.

### Data Analysis Key Findings
*   **K-Fold Cross-Validation Performance**: The Logistic Regression model achieved a mean accuracy of 0.9780 with a standard deviation of 0.0120 across 5 folds.
*   **Bootstrap Sampling Performance**: The model achieved a mean accuracy of 0.9776 with a standard deviation of 0.0069 over 100 bootstrap iterations.
*   **Accuracy Consistency**: Bootstrap Sampling provided a more stable estimate of accuracy, as indicated by its lower standard deviation (0.0069) compared to K-Fold Cross-Validation (0.0120).
*   **Method Advantages**:
    *   K-Fold Cross-Validation efficiently uses all data points for both training and testing, reducing bias and variance compared to a single train-test split.
    *   Bootstrap Sampling is robust for smaller datasets and can provide a distribution of performance estimates, allowing for confidence interval calculations.
*   **Method Disadvantages**:
    *   K-Fold Cross-Validation can still have higher variance for very small datasets, and the choice of 'k' can impact results.
    *   Bootstrap Sampling introduces an artificial bias due to sampling with replacement and can be more computationally intensive.

### Insights or Next Steps
*   Both evaluation methods confirm the high average performance of the Logistic Regression model on this dataset. The slightly lower variability observed with Bootstrap Sampling suggests it might be preferred when a more stable estimate of performance across different data subsets is critical.
*   To further validate the model's robustness, consider investigating the out-of-bag error from Bootstrap Sampling, which can serve as an unbiased estimate of the generalization error without needing a separate test set.
